### PR TITLE
Make the CC license field component configurable in DSpace 8.0

### PR DESCRIPTION
**Pull request by abelgomez**
## References
_Add references/links to any related issues or PRs. These may include:_
This PR does not directly address an open issue, but improves PR #3010 that addressed #3002.

## Description
The `ItemPageCcLicenseFieldComponent`implemented in #3010 has some configuration properties that, however, are hardcoded in the component itself and require manual modifications of the typescript code.

Since at least two configuration properties are indeed easily configurable in the back end (license URI field and license name field), makes sense that the front-end allows to tweak them.

This PR adds these configuration options (and some others) to the application config, so that no more modifications are required at the code level.

## Instructions for Reviewers

1. Run the front-end with the default DSpace 8.0 configuration.
2. Create an item, and set its `dc.rights.uri` and `dc.rights` to a valid value (e.g., `https://creativecommons.org/licenses/by-nc-nd/4.0/` and `CC BY-NC-ND 4.0` respectively) and see the Creative Commons badge to appear.
3. Delete the `dc.rights` metadata value and refresh the item. The CC badge will not show (the `ItemPageCcLicenseFieldComponent` requires both the CC uri and CC name to be set).
4. Create a new `dc.rights.license` metadata value with the value `CC BY-NC-ND 4.0`. Refresh the item, and the badge will still not be shown (since only `dc.right.uri` will be found with the default configuration).
5. Shutdown the front-end, and add the following configuration to your app config:
```
ccLicense:
  # Field name containing the CC license name, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license'
  # Defaults to 'dc.rights'
  nameField: dc.rights.license
```
6. Start the front-end, and the CC badge will now appear in the item.

Other properties that ca be configured are:

```
# Creative Commons metadata fields
ccLicense:
  # Icon variant:
  # 'full' variant shows image, a disclaimer (optional) and name (always), better for the item page content.
  # 'small' (default) variant shows image and name (optional), better for the item page sidebar
  variant: small
  # Field name containing the CC license URI, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license.uri'
  # Defaults to 'dc.rights.uri'
  uriField: dc.rights.uri
  # Field name containing the CC license name, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license'
  # Defaults to 'dc.rights'
  nameField: dc.rights
  # Shows the CC license name with the image. Always show if image fails to load
  showName: true
  # Shows the disclaimer in the 'full' variant of the component
  showDisclaimer: true
```

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `yarn lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `yarn check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [ ] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by selensoftr** [comment]
Not working in Publication, only working item.

**Comment by abelgomez** [comment]
The [`publication.component.html`](https://github.com/DSpace/dspace-angular/blob/main/src/app/item-page/simple/item-types/publication/publication.component.html) (the component that renders publications) does not use the `ds-item-page-cc-license-field`. Thus, it is not related to this PR.

`ds-item-page-cc-license-field` is, currently, only used by `https://github.com/DSpace/dspace-angular/blob/main/src/app/item-page/simple/item-types/publication/publication.component.html`.

IMHO, whether publications should (or should not) use `ds-item-page-cc-license-field` must be discussed in a different issue (or PR).

**Comment by abelgomez** [comment]
Thinking about the licensing of untyped items and publications... maybe you're right, and it should be added to the default templates. To be fair, it seems a bit arbitrary that it is only shown for untyped items and only when selecting a CC license.

Feel free to post a new issue. If the community thinks it's worth implementing, I volunteer to do so.

**Comment by selensoftr** [comment]
Dear Abelgomez, thank you very much for reply.

I added the relevant values ​​in the untyped-item.component.html and .ts file to the publication.component.hmtl and .ts file and ran the build, but it resulted in an error. (not: version DSpace 8.0)

**publication.component.ts**

import { ItemPageCcLicenseFieldComponent } from '../../field-components/specific-field/cc-license/item-page-cc-license-field.component';

@Component({
  selector: 'ds-publication',
  styleUrls: ['./publication.component.scss'],
  templateUrl: './publication.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: true,
  imports: [
ItemPageCcLicenseFieldComponent,
....

**publication.component.html**
<ds-item-page-cc-license-field [item]="object" [variant]="'full'"></ds-item-page-cc-license-field>

**yarn start:dev**

**Result:**
Error: src/app/item-page/simple/item-types/publication/publication.component.html:104:52 - error NG8002: Can't bind to 'variant' since it isn't a known property of 'ds-item-page-cc-license-field'.
1. If 'ds-item-page-cc-license-field' is an Angular component and it has 'variant' input, then verify that it is included in the '@Component.imports' of this component.
2. If 'ds-item-page-cc-license-field' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@Component.schemas' of this component to suppress this message.
3. To allow any property add 'NO_ERRORS_SCHEMA' to the '@Component.schemas' of this component.

104     <ds-item-page-cc-license-field [item]="object" [variant]="'full'">
                                                       ~~~~~~~~~~~~~~~~~~
  src/themes/custom/app/item-page/simple/item-types/publication/publication.component.ts:42:16
    42   templateUrl: '../../../../../../../app/item-page/simple/item-types/publication/publication.component.html',
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Error occurs in the template of component PublicationComponent.

**Comment by abelgomez** [comment]
Again, the error you have posted is unrelated to this PR (you would get the same error with the unmodified DSpace 8.0 code base).

It is caused because the file `src/themes/custom/app/item-page/simple/item-types/publication/publication.component.ts` references the template at `src/app/item-page/simple/item-types/publication/publication.component.html`.

Thus, you must also add `ItemPageCcLicenseFieldComponent` to the component imports in `src/themes/custom/app/item-page/simple/item-types/publication/publication.component.ts` as you did in `src/app/item-page/simple/item-types/publication/publication.component.ts`.

It also took me a couple of minutes to notice that the error was coming from the "custom" theme, rather than from the "base" app 😉

**Comment by selensoftr** [comment]
Dear @abelgomez thank you very much.
I did as you said and it worked.
Thank you very much.

File: config/config.example.yml [hunk 1]
```diff
@@ -502,7 +502,22 @@ notifyMetrics:
...
+# Creative Commons metadata fields
+ccLicense:
+  # Icon variant:
+  # 'full' variant shows image, a disclaimer (optional) and name (always), better for the item page content.
+  # 'small' (default) variant shows image and name (optional), better for the item page sidebar
+  variant: small
+  # Field name containing the CC license URI, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license.uri'
+  # Defaults to 'dc.rights.uri'
+  uriField: dc.rights.uri
+  # Field name containing the CC license name, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license'
```
Comment (tdonohue) [inline config/config.example.yml:514]: This line should say `cc.license.name` as that's the property in dspace.cfg that it should match

File: config/config.example.yml [hunk 2]
```diff
@@ -502,7 +502,22 @@ notifyMetrics:
...
-
+# Creative Commons metadata fields
+ccLicense:
+  # Icon variant:
+  # 'full' variant shows image, a disclaimer (optional) and name (always), better for the item page content.
+  # 'small' (default) variant shows image and name (optional), better for the item page sidebar
+  variant: small
+  # Field name containing the CC license URI, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license.uri'
+  # Defaults to 'dc.rights.uri'
+  uriField: dc.rights.uri
```
Comment (tdonohue) [inline config/config.example.yml:513]: Because this setting should always match the `cc.license.uri` setting in `dspace.cfg`, it'd be **better** to just have the backend share that setting with the user interface.  This is what we do in other areas of the application as well.

For example, the backend already does this with `cc.license.jurisdiction` setting in `dspace.cfg`.  In order to tell the backend to make that setting available, you just need to add it to the list of `rest.properties.exposed` in `rest.cfg` like this: https://github.com/DSpace/DSpace/blob/main/dspace/config/modules/rest.cfg#L54

Then, the UI code is able to **request** that setting from the REST API, like this:
https://github.com/DSpace/dspace-angular/blob/main/src/app/submission/sections/cc-license/submission-section-cc-licenses.component.ts#L310

So, this PR could be simplified if you removed the `uriField` setting here, and instead made a call to:
```
 this.configService.findByPropertyName('cc.license.uri')
```

In general, this PR does seem like a step in the right direction.  So, if you feel you cannot update it to read these values from the REST API, we still can move this forward.  But, I wanted to point out how easy it is to just read these settings from the REST API instead of duplicating them to the UI configs.

File: config/config.example.yml [hunk 3]
```diff
@@ -502,7 +502,22 @@ notifyMetrics:
...
+  # Field name containing the CC license name, as configured in the back-end, in the 'dspace.cfg' file, property 'cc.license'
+  # Defaults to 'dc.rights'
+  nameField: dc.rights
```
Comment (tdonohue) [inline config/config.example.yml:516]: Similar to the note above, we also could remove this config and instead configure the REST API to expose `cc.license.name` and then call ``
```
this.configService.findByPropertyName('cc.license.name')
```
Again, your approach here will work too...but it would require users to make sure these two settings are duplicated into both dspace.cfg and config.yml.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@abelgomez : Thanks for this work!  Apologies for the delayed response, but we did a team code review of this PR in today's Developer Meeting.  Overall, the feedback was positive & this PR seems like a nice improvement.  Just a few things we noticed:

1. First, this PR has code conflicts which need to be resolved so that it's mergeable again.
2. Second, the new configs are really useful. But, for `uriField` and `nameField`, you **could** just read these values from the REST API (if you enable them to be read as I describe below).  This would make this improvement even easier to use, as you can configure these fields in one place (on the backend) instead of in two places.

Overall though, only a rebase is required.  But, we'd still recommend considering reading these two settings from the backend as it will make this much easier to use.

**Comment by abelgomez** [comment]
Hi @tdonohue , thanks for the feedback. Reading the configuration from the backend seems the way to go.

Sorry for not noticing it could be done 🤦‍♂️🤦‍♂️ I still need to learn a lot about the codebase...

I'll work on the rebase, and I'll implement the improvements (since they also affect the implementation of other PRs I'm working on. Unfortunately, as I said some weeks ago in another PR, I'm a bit busy this season so that fixes will take some weeks.

Cheers!

**Comment by tdonohue** [comment]
@abelgomez : No worries on your timelines.  We expect that not all developers are able to get back to PRs immediately at all times.  But, whenever you are able to get back to this, please feel free to ping me for another (hopefully final) review.  Overall, the idea here is very good though & we appreciate your effort in making CC licenses more configurable.

**Comment by abelgomez** [comment]
Hi @tdonohue ,

I think that the code is ready. It now retrieves the `cc.license.uri` and `cc.license.name` from the backend if they are exposed.

Since these two properties do not hold any sensitive information, I'll open a PR in the backend to expose both of them by default.

I acknowledge that I have introduced a bit of confusion in the commits, but, in the end, the final code didn't change too much w.r.t. my previous proposal, so I hope it's easy to review.

The thing is that I initially removed all the configuration options (because I forgot about `variant`, `showName` and `showDisclaimer`), and then I realized that not only `uriField` and `nameField` were configurable, so I restored the part the dealt with  `variant`, `showName` and `showDisclaimer`.

I also initially decided to remove the `ccLicenseUriField` and `ccLicenseNameField` configurable properties from the component, but I later thought that it could be useful to override the backend configuration at the template level. Thus, I also decided to restore those properties.

Sorry for the indecisiveness.

On a side note, it's also noteworthy to mention that I wrapped part of the properties inside Observables, since now their values depend on the back-end configuration that is retrieved when the component is initialized.

Now, since the URI and license name are retrieved by the backend, the initial description and title of the PR are not fully accurate.  

Here is an updated version:

## Description

The `ItemPageCcLicenseFieldComponent`implemented in #3010 has some configuration properties that, however, are hardcoded in the component itself and require manual modifications of the typescript code.

Two configuration properties are configurable in the back end (license URI field and license name field), and other are only specific to the front-end (icon variant, show name, and show disclaimer).

This PR:
1. Retrieves the license URI field and license name field from the back-end (if the properties are exposed).
2. Adds configuration options for the icon variant, name visibility , and disclaimer visibility to the application config, so that no more modifications are required at the code level.

In any case, the component can still be used in angular templates as before. If a developer explicitly sets the `ccLicenseUriField` and `ccLicenseNameField` properties when using it, the configuration of the back-end will be ignored and the issued values will be used instead.

## Instructions for Reviewers

1. Run the front-end with the default DSpace 8.0 configuration.
4. Create an item, and set its `dc.rights.uri` and `dc.rights` to a valid value (e.g., `https://creativecommons.org/licenses/by-nc-nd/4.0/` and `CC BY-NC-ND 4.0` respectively) and see the Creative Commons badge to appear.
5. Delete the `dc.rights` metadata value and refresh the item. The CC badge will not show (the `ItemPageCcLicenseFieldComponent` requires both the CC uri and CC name to be set).
6. Create a new `dc.rights.license` metadata value with the value `CC BY-NC-ND 4.0`. Refresh the item, and the badge will still not be shown (since only `dc.right.uri` will be found with the default configuration).
7. **Shutdown the back-end**, and add the following configuration to your local.cfg:
8. 
```
rest.properties.exposed = cc.license.name
cc.license.name = dc.rights.license
```
9. Start the back-end, and the CC badge will now appear in the item.

The properties that may be read from the back-end are:

```
cc.license.name = dc.rights.license
cc.license.uri = dc.rights.uri
# In order to be exposed via the REST API, the following properties are also required:
rest.properties.exposed = cc.license.name
rest.properties.exposed = cc.license.uri
```

The properties that can be configured in the front-end are:

```
# Creative Commons metadata fields
ccLicense:
  # Icon variant:
  # 'full' variant shows image, a disclaimer (optional) and name (always), better for the item page content.
  # 'small' (default) variant shows image and name (optional), better for the item page sidebar
  variant: small
  # Shows the CC license name with the image. Always show if image fails to load
  showName: true
  # Shows the disclaimer in the 'full' variant of the component
  showDisclaimer: true
```

**Review by tdonohue — COMMENTED** [review]
@abelgomez : I'm trying to test this today, and I cannot seem to get it to work.  *Without* this PR, I can see the "Creative Commons license" section as normal.

cc-license

But, if I install this PR, and run it in production mode (`npm run build:prod; npm run serve:ssr`) and go to the **same Item**, then that section is gone.

no-cc-license

In both cases, I have an identical backend.  I'm not seeing any javascript errors in my browser's DevTools console.

Any ideas what could be going on?  Does this work for you in production mode?

**Comment by abelgomez** [comment]
Hi @tdonohue ,

I've been unable to replicate your issue.

I've run both the latest code in `main` and in this PR, and in both cases the CC license badge was shown when running in production mode. I tested this item: http://localhost:4000/items/d7191e7e-c849-4fed-99f3-1369b0c3e68e (I used sandbox.dspace.org as the backend). Of course, I executed `npm run clean` and `npm install` between executions.

However, when running this PR's code, I noticed that the CC field component took a few milliseconds more than the rest of the page to be rendered for the first time. However, even after clearing the cache, I've been unable to replicate this slight delay in subsequent calls/refreshes.

In any case, this small first delay makes sense: as opposed to other items' properties, the CC field component needs to make an additional call to the backend to retrieve the `cc.license.name` and `cc.license.uri` configuration properties to determine whether the item's license is CC or not (and thus, whether the component must be rendered or not).

Could it be the case that, for whatever reason, when you tested the PR, the query to the configuration backend was delayed, blocked, or didn't return a value yet (and thus, the component was still "deciding" if it had to be rendered or not)?

**Review by tdonohue — CHANGES_REQUESTED** [review]
@abelgomez : Thanks for verifying things on your end.  I tried again today, and I've figured out what is going on.  The problem is that it appears there's an accidental typo in your testing instructions in this PR.

Today, I've been able to verify that I *can* get this to work properly if I use this PR and https://github.com/DSpace/DSpace/pull/9882 with their default settings.  In that scenario, I see the CC license information on the Item Page.

However, previously, I was setting this in my `local.cfg` (which causes the CC License to not be displayed at all): 
```
cc.license.name = dc.rights.license
cc.license.uri = dc.rights.uri
```

I had copied & pasted this from your instructions without looking at it closely.  But, I now realize it has a typo which caused it not to work.  It should be this:
```
cc.license.name = dc.rights
cc.license.uri = dc.rights.uri
```

So, testing again today, I can verify that this **is working properly**.  It displays the license when no `cc.license.*` values are set in local.cfg.  It also displays the license when they are set to correct `dc.rights.*` values.  It does *not* display the CC license if the specified `dc.rights.*` fields do not exist (which is what I was accidently testing before).

I'm going to do one last review of the code (and possibly merge this later today if it looks good).  I can also fix the accidental typo in the testing instructions for this PR.

**Comment by tdonohue** [comment]
(Whoops, I accidently submitted that last review as a "request for changes".  That can be disregarded.  I'll re-review the code in the next few hours and submit a final review.)

**Comment by abelgomez** [comment]
Ooops!

That's because I copied and pasted those keys from my `local.cfg`, and I had modified them to test that the CC badge was working properly when non-default dc keys were being used for CC licenses (which is the case for our organization, which uses `dc.rights.license` instead of `dc.rights`).

Thanks for double-checking, and sorry for my mistake.

On a side note, merging this PR may introduce conflicts in https://github.com/DSpace/dspace-angular/pull/3338, since both modify the same methods in `ItemPageCcLicenseFieldComponent`. Thus, I think that it would be better to **merge this PR first, and merge https://github.com/DSpace/dspace-angular/pull/3338 afterwards**. If conflicts are introduced, I'll fix them ASAP.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@abelgomez : I gave this a code review & some more testing of the settings.  Overall, it is working. However, I've found what looks to be a possible bug.

If I try to just use the "small" image with no name or disclaimer, then I get no license displayed.  So, I expected this would _only display the CC license icon_:
```
ccLicense:
  variant: small
  showName: false
  showDisclaimer: false
```
I was just surprised that there doesn't seem to be a way to turn off both the name and disclaimer..you seem to only be able to turn off *one* of them.

Because the CC license is ONLY used on the Item page, I also wonder if we should consider embedding these configurations under the existing `item` configuration like this:
```
item:
  ccLicense:
    variant: small
    showName: true
    showDisclaimer: true
```
To me, this might make more sense because the `ccLicense` configuration ONLY works for the Item page.  However, I wouldn't require this if you feel that it might be used elsewhere eventually.

Other than the above bug, this looks to be working well.

**Comment by abelgomez** [comment]
Hi @tdonohue ,

you are right, with such configuration keys, the badge does not show which is a bug.

Nevertheless, checking what was the original behavior (contributed in https://github.com/DSpace/dspace-angular/pull/3010 by another dev), it also seems kind of broken, since the `showDisclaimer` should only be considered in the full variant, not the small. Thus, in the original implementation, it was not considered the case of showing only the small badge without any kind of text.

I can rework the logic to show the badge to consider such cases if you think it's worth it. Not sure I'll be able to do it during the weekend.

Regarding the second point, I think we should keep the `ccLicense` config key as a "top-level" one. This is foreseeing that https://github.com/DSpace/dspace-angular/pull/3338 will be merged. In that PR, the CC license badge is also shown in "Publication" entity types, fixing https://github.com/DSpace/dspace-angular/issues/3167.

**Comment by abelgomez** [comment]
Hi @tdonohue , I agree with both points.

I'll fix the bug I introduced that under some circumstances does not show the badge, and I'll try to fix the original behavior to make it more consistent.

Also, if configs under `item` apply to all entity types, I agree that maybe it's better to place them under the item key. I'll also work on that. I think that tomorrow I'll have time to work on both points.

**Comment by abelgomez** [comment]
Hi @tdonohue ,

I finished reworking this PR. I made some significant changes with respect to the original commit to make the configuration options more understandable and consistent.

1. I replaced the `variant` configuration key with `position`. This is because the `variant` property of the `ItemPageCcLicenseFieldComponent` component is more of an internal configuration option that changes the layout to adapt the component to its context rather than to configure the actual size. The `small` variant renders the different parts of the license stacked one on top of the other (and thus, it is designed to be used in the sidebar), while the `full` variant renders the license info in columns (targeted to be used in the contents part of the item page). This is something that developers reusing this component must set in their own templates, rather than end-users using the default templates should configure.

   Thus, in the `item > ccLicense` config keys I no longer expose `variant` (which is an implementation detail), but I include a `position` configuration key that is targeted to the end-users. The `position` allows setting whether the license information should be placed in the `sidebar` or in the `contents` part of an item. I updated the untyped item template accordingly.

2. Now, the license badge can be rendered without any text (i.e., both `showName` and `showDisclaimer` can be false). The license text can be hidden either if the license is rendered in the sidebar or the contents area. Conversely, the license text can be shown (with or without the disclaimer text) both in the sidebar or in the contents area (in the original implementation, the disclaimer was always hidden in the `small` variant, now is up to the users or developers).

3. Now, the `variant` argument of the `ItemPageCcLicenseFieldComponent` is mandatory, since developers reusing this component in their own templates have to take into account whether the component is rendered in the sidebar or the contents area. This consideration has not changed w.r.t. the original implementation, but we are now making it explicit to avoid confusion about the actual behavior of the `variant` parameter. I think it is fair to change the "optionality" of this property since DSpace 9 already includes important breaking changes w.r.t. DSpace 8 in the angular UI (notably the https://github.com/DSpace/dspace-angular/pull/3997 PR) which will require developers to reconcile their manually modified templates with DSpace 9.

6. The configuration options are now documented as follows in `config.example.yml`:

```
# Item Config
item:
  # Creative Commons metadata fields
  ccLicense:
    # Icon variant:
    # 'content' (default) variant shows license in the item page content.
    # 'sidebar' variant shows license in the item page sidebar
    position: content
    # Shows the CC license name with the image. The name is always shown if the image fails to load
    showName: true
    # If the CC license name is shown, also show a disclaimer text. Only recommendend in the 'content' position
    # If showName is set to false, this configuration key does not have any effect
    showDisclaimer: true
```

Let me know if this PR looks good now, so I can start reworking https://github.com/DSpace/dspace-angular/pull/3338 accordingly.

**Comment by alexandrevryghem** [comment]
I haven't really tested this PR, but I don't think that we should implement an option to configure on which side of the item page the license should be displayed, because if we do this for all the fields this is going to become very messy. Often when institutions want to change the position of the fields, they also want to change the order of the fields inside the left/right column and also request additional field to be displayed well. In such cases they would already need to theme the item page so such changes need to happen in the src/themes instead

**Comment by abelgomez** [comment]
The point of allowing setting it in one place or the other is because, for me, if users want to show only the badge (or the badge and a short link), it makes more sense to show it in the sidebar. But if all the info is shown (the badge, the license text, and the disclaimer), it makes more sense to show it in the contents area.

The whole point of this PR is to make easily configurable which info is shown for those who don't want/know how to customize the templates, and thus, it made sense for me to allow configuring the actual place.

If users have the knowledge to create their own templates, they can use the `ItemPageCcLicenseFieldComponent` as before, setting the configuration of the component as properties when using it in their own template.

**Comment by abelgomez** [comment]
Just to complement my previous comment: if allowing to set the position of the CC badge is too overkill, I think that allowing to configure whether the license name or the disclaimer should be shown is almost useless (as I said, I think that the amount of information to be shown depends on the place of the badge).

Thus, in such a case, I vote for removing all configuration options and always show the full badge, with the disclaimer, in the contents area. In the end, the idea of this PR started to allow setting the `ccLicenseUriField` and `ccLicenseNameField`properties, which were hardcoded. If a DSpace instance was using other DC fields that the defaults, manual changes were required either in the untyped item template or in the `ItemPageCcLicenseFieldComponent`.

Thus, since I allowed to set those properties, I thought that adding a few more properties would not harm and add some flexibility. Then, at the beginning, I implemented `ccLicenseUriField`, `ccLicenseNameField` as configuration properties (together with the ones that remain), but as suggested by @tdonohue (and related to PR https://github.com/DSpace/DSpace/pull/9882), they can be retrieved from the backend, and thus no configuration properties are needed.

In conclusion, if setting the look and feel of the badge is too much, I vote for removing all the config properties and limit this PR to retrieve the URI and name fields of the license from the backend.

Let me know your thoughts. Cheers!

**Comment by alexandrevryghem** [comment]
@abelgomez: It’s not because we make something not configurable through the yml files, that it shouldn’t be configurable at all. Most of the config options are currently only configurable with `@Input`s, especially for item page fields I think it’s more logical to make everything configurable on component level instead of a side wide option. The options will still be able to have a default and they would still be configurable, but this would provide even greater customizability, as it would allow institutions to customize their Publication page in a different way than their OrgUnit page for example.

**Comment by tdonohue** [comment]
@abelgomez : I'm finally finding time to revisit this PR today.  I agree with @alexandrevryghem that I'm not a fan of the newly added `position` configuration, because it may be verify difficult to keep that working if a site chooses to modify the Item page (`untyped-item.component.html`) in their theme.  

Usually, we allow sites to move components around on the page by just copying & pasting the component HTML tag to the new location.  So, technically, a site can already move this component via the HTML.  This configuration goes against that approach.

I'm also seeing your point that this may make the `showName` and `showDisclaimer` options less useful.  Maybe it's better to just **simplify** and let a site choose the `variant` (small or full) in the `untyped-item.component.html`?  Maybe that's all the configuration we really need.  As you noted, most sites may want the `full` view if they keep the license display as-is (in the body of the page), but may choose to switch it to `small` if they want to move it to the sidebar area.

**Comment by tdonohue** [comment]
@abelgomez : I see your point that the `variant` itself can just be managed via HTML templates.

I'm OK with your suggestion to just remove the configurations from the `config.*.yml` file and revert this PR to only getting the configs from the backend (which would require merging https://github.com/DSpace/DSpace/pull/9882 -- but that PR seems non-controversial).

**Review by alexandrevryghem — COMMENTED** [review]
@abelgomez: Thnx I looked again at this today since it's related to #3338. I added some inline comments, could you take a look at them when you have some time

**Comment by abelgomez** [comment]
Hi @alexandrevryghem ,

Thanks for the feedback! I'm already working on your comments... Cheers!

File: src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.html [hunk 4]
```diff
@@ -1,32 +1,33 @@
-@if (uri && name) {
+@if ((uri$ | async) && (name$ | async)) {
```
[the remark this answers is not in the document]
Comment (abelgomez) [reply src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.html]: Done!

File: src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.ts [hunk 5]
```diff
@@ -33,16 +47,14 @@ export class ItemPageCcLicenseFieldComponent implements OnInit {
...
-  @Input() ccLicenseUriField? = 'dc.rights.uri';
+  @Input() ccLicenseUriField?;
 
   /**
-   * Filed name containing the CC license name, as configured in the back-end, in the 'dspace.cfg' file, property
-   * 'cc.license.name'
+   * Field name containing the CC license name
    */
-  @Input() ccLicenseNameField? = 'dc.rights';
+  @Input() ccLicenseNameField?;
```
[the remark this answers is not in the document]
Comment (abelgomez) [reply src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.ts]: Done!

File: src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.ts [hunk 6]
```diff
@@ -54,19 +66,60 @@ export class ItemPageCcLicenseFieldComponent implements OnInit {
...
+      getFirstCompletedRemoteData(),
+      getRemoteDataPayload(),
+    ).subscribe((remoteData: ConfigurationProperty) => {
+      if (this.ccLicenseNameField === undefined) {
+        // Set the value only if it has not manually set when declaring this component
+        this.ccLicenseNameField = remoteData?.values && remoteData?.values?.length > 0 ? remoteData.values[0] : 'dc.rights';
+      }
+      this.name$ = of(this.item.firstMetadataValue(this.ccLicenseNameField));
+    }),
+    );
```
[the remark this answers is not in the document]
Comment (abelgomez) [reply src/app/item-page/simple/field-components/specific-field/cc-license/item-page-cc-license-field.component.ts]: I have refactored the code to not use manual subscribes, but I left the initialization code in the ngOnInit method. To be fair, I was not able to make the code work with the `switchMap` when the `ccLicenseNameField` property was initialized directly in the "calling" template.

I hope the current implementation matches your standards 😊

File: src/app/shared/utils/license.utils.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+/**
+   * Returns whether a URI denotes a valid URI for CC licenses.
+   * @param uri
+   * @returns true if the URI corresponds to a reconigzable CC license URI or false otherwise
+   */
+export function isCcLicense(uri: string): boolean {
+  return !!parseCcCode(uri);
+}
```
[the remark this answers is not in the document]
Comment (abelgomez) [reply src/app/shared/utils/license.utils.ts:19]: Ops!

I copied the utility class as it is in  PR #3338, where it is used. I'll remove it.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @abelgomez !  I gave this another look/test today, and it's working properly. I've also verified I can modify `cc.license.name` and `cc.license.uri` on the backend in order to get the license information to display from a different metadata field.
